### PR TITLE
LSIF: Improve remote references in same repo

### PR DESCRIPTION
**Pull request by efritz**
This implements [RFC 58](https://docs.google.com/document/d/1Ccp89K_Yxq6yesBmUvQFhlczcP-PQxb7-HsmqNnqbLs) which will special-case finding remote references for dumps of the same repository.

For dumps of other repositories, these dumps must be visible at the tip. For dumps of the same repository, the closest commits are chosen with non-overlapping roots.

**Reviewer notes**:

Only 320 relevant lines changed: backend and xrepo. The rest are updates to testing and test utils.

Backend has split `remoteReferences` into `remoteReferences`, a similar function `sameRepositoryRemoteReferences`, and a common helper `locationsFromRemoteReferences`. These methods basically call an equivalent method in xrepo. This file also updated its pagination to deal with the case where we're looking at the result set consisting of _same_ repo dumps, or _distinct_ repo dumps.

Xrepo adds the method `getSameRepoRemoteReferences`.

**Tasklist**:
- [x] Add explicit test for *both* pagination conditions at once.

### CODE REVIEW

File: lsif/src/xrepo.ts [hunk 1]
```diff
@@ -513,7 +518,140 @@ export class XrepoDatabase {
...
+                const args = [scheme, name, version, visible_ids, offset, limit]
+                const results = await entityManager.query(referenceIdsQuery, args)
+                const referenceIds = extractIds(results)
+                const references = await entityManager.getRepository(ReferenceModel).findByIds(referenceIds)
+
+                // findByIds doesn't return models in the same order as they were requested,
+                // so we need to sort them here before returning. This sorts the models by
+                // the index of the model's id in `referenceIds`.
+                const indexes = new Map(referenceIds.map((v, i) => [v, i]))
+                return references.sort((a, b) => (indexes.get(a.id) || -1) - (indexes.get(b.id) || -1))
```
Comment (chrismwendt) [inline lsif/src/xrepo.ts]: Perhaps a more straightforward way to do this is:

```typescript
const referenceById = new Map(references.map(reference => [reference.id, reference]))
return referenceIds.map(id => referenceById[id])
```

File: lsif/src/xrepo.ts [hunk 2]
```diff
@@ -513,7 +518,140 @@ export class XrepoDatabase {
...
+            // as a parameter.
+            const visible_ids = extractIds(await entityManager.query(visibleIdsQuery, [repository, commit]))
+
+            // Get total number of items in this result set
+            const rawCount = await entityManager.query(countQuery, [scheme, name, version, visible_ids])
+            const totalCount = parseInt((rawCount as { count: string }[])[0].count, 10)
+
+            // Construct method to select a page of possible references. We first perform
+            // the query defined above that returns reference identifiers, then perform a
+            // second query to select the models by id so that we load the relationships.
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/xrepo.ts]: It is a join in app code :/

I think there's a way to do it, but it requires selecting all of the columns explicitly the same way the typeorm query would (which would be brittle and easy to add a column and break this). 

I'm thinking about looking at a different library for the SQL abstraction and this is definitely a data point against typeorm.

File: lsif/src/xrepo.ts [hunk 3]
```diff
@@ -496,10 +500,11 @@ export class XrepoDatabase {
                 .createQueryBuilder('reference')
                 .leftJoinAndSelect('reference.dump', 'dump')
                 .where({ scheme, name, version })
+                .andWhere('dump.repository != :repository', { repository })
                 .andWhere('dump.visible_at_tip = true')
 
-            // Get pagination count
-            const count = await baseQuery.getCount()
+            // Get total number of items in this result set
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/xrepo.ts]: No, it's just the sql result set. Is there a way to make this more clear?

[the remark this answers is not in the document]
Comment (chrismwendt) [reply lsif/src/xrepo.ts]: "set of results" would at least break the ambiguity with LSIF's concept of "result set".

[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/xrepo.ts]: Updated

**Pull request merged**